File: Semestr 3/Paradygmaty Programowania - laboratorium/lista3/Scala/lista3_lab.scala
import scala.annotation.tailrec

@tailrec
def lastElement[A](list: List[A]): Option[A] = {
  list match {
    case Nil => None
    case x :: Nil => Some(x)
    case _ :: tail => lastElement(tail)
  }
}

@tailrec
def twoLastElements[A](list: List[A]): Option[(A, A)] = {
  list match {
    case Nil => None
    case _ :: Nil => None
    case a :: b :: Nil => Some((a, b))
    case _ :: tail => twoLastElements(tail)
  }
}

def listLength[A](list: List[A]): Int = {
  list match {
    case Nil => 0
    case _ :: tail => 1 + listLength(tail)
  }
}

def reverseList[A](list: List[A]): List[A] = {
  list match {
    case Nil => Nil
    case head :: tail => reverseList(tail) :+ head
  }
}

def isPalindrome(s : String): Boolean = {
  @tailrec
  def isPalindromeHelper(s: String): Boolean = {
    if (s.length <= 1) {
      true
    } else if (s.head != s.last) {
      false
    } else {
      isPalindromeHelper(s.substring(1, s.length - 1))
    }
  }
  
  val cleanString = s.replaceAll("\\s", "").toLowerCase()
  isPalindromeHelper(cleanString)
}

def removeDuplicates[A](list: List[A]): List[A] = {
  def removeAllOccurences(list: List[A], element: A): List[A] = {
    list match {
      case Nil => Nil
      case head :: tail if head == element =>
        removeAllOccurences(tail, element)
      case head :: tail =>
        head :: removeAllOccurences(tail, element)
    }
  }

  list match {
    case Nil => Nil
    case head :: tail =>
      val filteredTail = removeAllOccurences(tail, head)
      head :: removeDuplicates(filteredTail)
  }
}

def onlyEvenIndexes[A](list: List[A]): List[A] = {
  def onlyEvenIndexesHelper(list: List[A], index: Int): List[A] = {
    list match {
      case Nil => Nil
      case head :: tail if index % 2 == 0 =>
        head :: onlyEvenIndexesHelper(tail, index + 1)
      case head :: tail =>
        onlyEvenIndexesHelper(tail, index + 1)
    }
  }

  onlyEvenIndexesHelper(list, 0)
}

def isPrime(n: Int): Boolean = {
  @tailrec
  def isPrimeHelper(n: Int, divisor: Int): Boolean = {
    if (divisor * divisor > n) {
      true
    } else if (n % divisor == 0) {
      false
    } else {
      isPrimeHelper(n, divisor + 2)
    }
  }

  if (n <= 1){
    false
  } else if (n <= 3){
    true
  } else if (n % 2 == 0){
    false
  } else{
    isPrimeHelper(n, 3)
  }
}


@main
def main(): Unit = {
  val list1 = List(4, 1, 4, 2, 3, 4, 5, 4, 3, 2, 2, 4)
  val list2 = List()
  val list3 = List(5)
  val list4 = List(7, 7)

  println("\nLast element:")
  println(lastElement(list1))
  println(lastElement(list2))
  println(lastElement(list3))
  println(lastElement(list4))

  println("\nTwo last elements:")
  println(twoLastElements(list1))
  println(twoLastElements(list2))
  println(twoLastElements(list3))
  println(twoLastElements(list4))

  println("\nList length:")
  println(listLength(list1))
  println(listLength(list2))
  println(listLength(list3))
  println(listLength(list4))

  println("\nReverse list:")
  println(reverseList(list1))
  println(reverseList(list2))
  println(reverseList(list3))
  println(reverseList(list4))

  println("\nIs palindrome:")
  println(isPalindrome("ka mil slima   k"))
  println(isPalindrome("test"))

  println("\nRemove duplicates:")
  println(removeDuplicates(list1))
  println(removeDuplicates(list2))
  println(removeDuplicates(list3))
  println(removeDuplicates(list4))

  println("\nOnly even indexes:")
  println(onlyEvenIndexes(list1))
  println(onlyEvenIndexes(list2))
  println(onlyEvenIndexes(list3))
  println(onlyEvenIndexes(list4))

  println("\nIs prime:")
  println(isPrime(47))
  println(isPrime(5))
  println(isPrime(4))
  println(isPrime(25))
  println(isPrime(15))
}

File: Semestr 3/Paradygmaty Programowania/laboratorium/lista03/OCaml/lista3_lab.ml
let rec lastElement list =
  match list with
  | [] -> None
  | [x] -> Some x
  | _ :: tail -> lastElement tail



let rec twoLastElements list =
  match list with
  | [] -> None
  | [_] -> None
  | [a; b] -> Some (a, b)
  | _ :: tail -> twoLastElements tail



let rec listLength list =
  match list with
  | [] -> 0
  | _ :: tail -> 1 + listLength tail



let rec reverseList list = 
  match list with
  | [] -> []
  | head :: tail -> reverseList tail @ [head]



let isPalindrome s =
  let rec isPalindromeHelper s =
    if String.length s <= 1 then
      true
    else if s.[0] <> s.[(String.length s) - 1] then
      false
    else
      isPalindromeHelper (String.sub s 1 ((String.length s) - 2))
  in
  let clean_string = String.lowercase_ascii (String.concat "" (String.split_on_char ' ' s)) in
  isPalindromeHelper clean_string



let rec removeDuplicates list = 
  let rec removeAllOcurrences list element = 
    match list with
    | [] -> []
    | head :: tail when head = element -> 
      removeAllOcurrences tail element
    | head :: tail -> head :: removeAllOcurrences tail element
  in
  match list with
  | [] -> []
  | head :: tail -> 
    let filteredTail = removeAllOcurrences tail head in
    head :: removeDuplicates filteredTail



let onlyEvenIndexes list =
  let rec onlyEvenIndexesHelper list index = 
    match list with
    | [] -> []
    | head :: tail when index mod 2 = 0 ->
      head :: onlyEvenIndexesHelper tail (index+1)
    | head :: tail ->
      onlyEvenIndexesHelper tail (index+1)
  in
  onlyEvenIndexesHelper list 0



let isPrime n =
  let rec isPrimeHelper n divisor = 
    if divisor*divisor > n then
      true
    else if n mod divisor = 0 then
      false
    else
      isPrimeHelper n (divisor+2)
  in
  if n <= 1 then
    false
  else if n <= 3 then
    true
  else if n mod 2 = 0 then
    false
  else
    isPrimeHelper n 3

    

let () =
  let list1 = [4; 1; 4; 2; 3; 4; 5; 4; 3; 2; 2; 4] in
  let list2 = [] in
  let list3 = [5] in
  let list4 = [7; 7] in

  Printf.printf "\nLast element:\n";
  Printf.printf "%s\n" (match lastElement list1 with Some x -> string_of_int x | None -> "None");
  Printf.printf "%s\n" (match lastElement list2 with Some x -> string_of_int x | None -> "None");
  Printf.printf "%s\n" (match lastElement list3 with Some x -> string_of_int x | None -> "None");
  Printf.printf "%s\n" (match lastElement list4 with Some x -> string_of_int x | None -> "None");

  Printf.printf "\nTwo last elements:\n";
  Printf.printf "%s\n" (match twoLastElements list1 with Some (a, b) -> Printf.sprintf "(%d, %d)" a b | None -> "None");
  Printf.printf "%s\n" (match twoLastElements list2 with Some (a, b) -> Printf.sprintf "(%d, %d)" a b | None -> "None");
  Printf.printf "%s\n" (match twoLastElements list3 with Some (a, b) -> Printf.sprintf "(%d, %d)" a b | None -> "None");
  Printf.printf "%s\n" (match twoLastElements list4 with Some (a, b) -> Printf.sprintf "(%d, %d)" a b | None -> "None");

  Printf.printf "\nList length:\n";
  Printf.printf "%d\n" (listLength list1);
  Printf.printf "%d\n" (listLength list2);
  Printf.printf "%d\n" (listLength list3);
  Printf.printf "%d\n" (listLength list4);

  Printf.printf "\nReverse list:\n";
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (reverseList list1)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (reverseList list2)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (reverseList list3)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (reverseList list4)));

  Printf.printf "\nIs palindrome:\n";
  Printf.printf "%b\n" (isPalindrome "ka mil slima   k");
  Printf.printf "%b\n" (isPalindrome "test");

  Printf.printf "\nRemove duplicates:\n";
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (removeDuplicates list1)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (removeDuplicates list2)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (removeDuplicates list3)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (removeDuplicates list4)));

  Printf.printf "\nOnly even indexes:\n";
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (onlyEvenIndexes list1)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (onlyEvenIndexes list2)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (onlyEvenIndexes list3)));
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int (onlyEvenIndexes list4)));

  Printf.printf "\nIs prime:\n";
  Printf.printf "%b\n" (isPrime 47);
  Printf.printf "%b\n" (isPrime 5);
  Printf.printf "%b\n" (isPrime 4);
  Printf.printf "%b\n" (isPrime 25);
  Printf.printf "%b\n" (isPrime 15)
;;